### PR TITLE
daemon,client,docs: send install options to daemon

### PR DESCRIPTION
**Pull request by niemeyer**
This also changes both the client and the daemon to
consistently and uniquely use mime multipart uploads.

### CODE REVIEW

File: client/snap_op.go [hunk 1]
```diff
@@ -20,87 +20,113 @@
...
+	"io"
+	"mime/multipart"
 	"os"
-	"strings"
+	"path/filepath"
+	"strconv"
 )
 
+type SnapOptions struct {
+	Channel string `json:"channel,omitempty"`
```
Comment (zyga) [inline client/snap_op.go]: It's not a big deal but this allows to sideload a snap *and* specify a channel which is meaningless. I don't mind this as I understand how it makes everything below much more streamlined.

File: client/snap_op.go [hunk 2]
```diff
@@ -20,87 +20,113 @@
...
-	path := fmt.Sprintf("/v2/snaps/%s", name)
-	body := strings.NewReader(`{"action":"remove"}`)
+// Refresh refreshes the snap with the given name (switching it to track
+// the given channel if given).
+func (client *Client) Refresh(name string, options *SnapOptions) (changeID string, err error) {
+	return client.doSnapAction("refresh", name, options)
+}
 
-	return client.doAsync("POST", path, nil, body)
+func (client *Client) doSnapAction(actionName string, snapName string, options *SnapOptions) (changeID string, err error) {
```
Comment (mvo5) [inline client/snap_op.go]: Nice, I like this.

File: client/snap_op.go [hunk 3]
```diff
@@ -20,87 +20,113 @@
...
+	}
 
-// PurgeSnap purges the snap with the given name, returning the UUID of the
-// background operation upon success.
-//
-// TODO: nuke purge, when we have snapshots/backups done
-func (client *Client) PurgeSnap(name string) (changeID string, err error) {
-	path := fmt.Sprintf("/v2/snaps/%s", name)
-	body := strings.NewReader(`{"action":"purge"}`)
+	pr, pw := io.Pipe()
```
Comment (mvo5) [inline client/snap_op.go]: Nice pattern

File: cmd/snap/cmd_snap_op.go [hunk 4]
```diff
@@ -81,13 +81,9 @@ func wait(client *client.Client, id string) error {
 }
 
 var (
-	shortInstallHelp    = i18n.G("Install a snap to the system")
```
Comment (mvo5) [inline cmd/snap/cmd_snap_op.go]: Good catch to remove the purge/rollback/activate/deactivate ones.

**Comment by mvo5** [comment]
Looks good

**Comment by niemeyer** [comment]
retest this please

File: docs/rest.md [hunk 5]
```diff
@@ -310,22 +305,16 @@ If present, only list snaps that match the query.
 
 ### POST
 
-* Description: Sideload a snap to the system.
+* Description: Install an uploaded snap to the system.
```
Comment (chipaca) [inline docs/rest.md]: "uploaded" here isn't really clear. Maybe something like "install a snap provided in the request body..."?

No, that's not much better. Hmm.

File: cmd/snap/cmd_snap_op_test.go [hunk 6]
```diff
@@ -39,6 +39,7 @@ func (s *SnapSuite) TestInstall(c *check.C) {
 			c.Check(r.URL.Path, check.Equals, "/v2/snaps/foo.bar")
 			c.Check(DecodedRequestBody(c, r), check.DeepEquals, map[string]interface{}{
 				"action":  "install",
+				"name":    "foo.bar",
```
Comment (chipaca) [inline cmd/snap/cmd_snap_op_test.go]: hmm, why `.bar`?

Comment (chipaca) [inline cmd/snap/cmd_snap_op_test.go]: ah. test could use updating...

Comment (chipaca) [inline cmd/snap/cmd_snap_op_test.go]: ah. test could use updating...

Comment (niemeyer) [inline cmd/snap/cmd_snap_op_test.go]: Just preserving the remaining of the tests untouched. We can fix it all in one swap next week.

File: docs/rest.md [hunk 7]
```diff
@@ -310,22 +305,16 @@ If present, only list snaps that match the query.
...
-has an `allow-unsigned` field (with any value), the snap may be unsigned;
-otherwise attempting to sideload an unsigned snap will result in a failed
-background operation.
-
-It's also possible to provide the snap as the entire body of a `POST` (not a
-multipart request). In this case the header `X-Allow-Unsigned` may be used to
-allow sideloading unsigned snaps.
+The snap to install must be provided as part of the body of a
+`mutlipart/form-data` request. The form should have one file
+named "snap".
```
Comment (chipaca) [inline docs/rest.md]: probably needs to talk about the rest of the form, here

Comment (chipaca) [inline docs/rest.md]: probably needs to talk about the rest of the form, here

**Comment by chipaca** [comment]
LGTM. A few comments but no blockers.

**Comment by niemeyer** [comment]
Thanks! Let's move it in then. We can do these tweaks next week with more time.

**Pull request merged**